### PR TITLE
feat(observability/alerts): Alerts support

### PR DESCRIPTION
**Pull request by jsonroy-fastly**

### CODE REVIEW

File: pkg/commands/alerts/create.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		},
+	}
+	c.CmdClause = parent.Command("create", "Create Alerts")
+	c.CmdClause.Flag("name", "Name of the alert definition").Required().StringVar(&c.name)
+	c.CmdClause.Flag("description", "Description of the alert definition").Required().StringVar(&c.description)
+	c.CmdClause.Flag("metric", "Metric to alert on").Required().StringVar(&c.metric)
+	c.CmdClause.Flag("source", "Source to get the metric from").Required().StringVar(&c.source)
+	c.CmdClause.Flag("type", "Type of evaluation strategy, one of [all_above_threshold, above_threshold, below_threshold, percent_absolute, percent_decrease, percent_increase]").Required().StringVar(&c.eType)
+	c.CmdClause.Flag("period", "Period of time to evaluate, one of [2m, 3m, 5m, 15m, 30m]").Required().StringVar(&c.period)
+	c.CmdClause.Flag("threshold", "Threshold use to compare during evaluation").Required().Float64Var(&c.threshold)
```
Comment (Integralist) [inline pkg/commands/alerts/create.go:27]: We should sort the flags alphabetically (same for the optional flags below).

> **NOTE:** This comment applies throughout this PR.

File: pkg/commands/alerts/create.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			Globals: g,
+		},
+	}
+	c.CmdClause = parent.Command("create", "Create Alerts")
+	c.CmdClause.Flag("name", "Name of the alert definition").Required().StringVar(&c.name)
+	c.CmdClause.Flag("description", "Description of the alert definition").Required().StringVar(&c.description)
+	c.CmdClause.Flag("metric", "Metric to alert on").Required().StringVar(&c.metric)
+	c.CmdClause.Flag("source", "Source to get the metric from").Required().StringVar(&c.source)
+	c.CmdClause.Flag("type", "Type of evaluation strategy, one of [all_above_threshold, above_threshold, below_threshold, percent_absolute, percent_decrease, percent_increase]").Required().StringVar(&c.eType)
+	c.CmdClause.Flag("period", "Period of time to evaluate, one of [2m, 3m, 5m, 15m, 30m]").Required().StringVar(&c.period)
```
Comment (Integralist) [inline pkg/commands/alerts/create.go:26]: Wherever explicit/predefined inputs are expected, use an `EnumVar` along with`.HintOptions`, see the following for an example:

https://github.com/fastly/cli/blob/main/pkg/commands/serviceauth/create.go#L40

File: pkg/commands/alerts/doc.go [hunk 3]
```diff
@@ -0,0 +1,2 @@
+// Package alerts contains commands to inspect the CLI configuration.
```
Comment (Integralist) [inline pkg/commands/alerts/doc.go:1]: This description sounds wrong.

I imagine it should be more like:

```go
// Package alert contains commands to inspect and manipulate Fastly Alerts.
```

File: pkg/commands/alerts/list.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	definitionsPtr := make([]*fastly.AlertDefinition, len(definitions.Data))
+	for i := range definitions.Data {
+		definitionsPtr[i] = &definitions.Data[i]
+	}
+
+	if c.Globals.Verbose() {
+		printVerbose(out, &definitions.Meta, definitionsPtr)
+	} else {
+		printSummary(out, &definitions.Meta, definitionsPtr)
+	}
```
Comment (Integralist) [inline pkg/commands/alerts/list.go:77]: I'm not sure how useful it is to display the cursor information.

A CLI user is going to expect the list command to return all data and not stop after one page of paginated data. 

So we should either automatically follow all cursors until we've read all the data or if that could be a memory concern, we should prompt the user to "request another page of data". You can see an example of this approach here...

https://github.com/fastly/cli/blob/6c1ec982475e698bd347c94622f857d9e4f44c90/pkg/commands/secretstoreentry/list.go#L63-L75

**Review by Integralist — COMMENTED** [review]
Looking good. Just a few comments to address.

File: pkg/commands/alerts/common.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+		threshold := definition.EvaluationStrategy["threshold"].(float64)
+		text.Indent(out, indent+4, "  Threshold: %v", fastly.ToValue(&threshold))
+
+		if ignoreBelow, ok := definition.EvaluationStrategy["ignore_below"].(float64); ok {
+			text.Indent(out, indent+4, "  IgnoreBelow: %v", fastly.ToValue(&ignoreBelow))
+		}
+
+		integrations := definition.IntegrationIDs
+		if ok && len(integrations) > 0 {
```
[the remark this answers is not in the document]
Comment (jsonroy-fastly) [reply pkg/commands/alerts/common.go:43]: Definitely a copy/paste mistake.

File: pkg/commands/alerts/common.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/fastly/cli/pkg/text"
+	"github.com/fastly/go-fastly/v9/fastly"
+)
+
+func printDefinition(out io.Writer, indent uint, definition *fastly.AlertDefinition) {
+	if definition != nil {
+		text.Indent(out, indent, "Definition ID: %s", fastly.ToValue(&definition.ID))
+		text.Indent(out, indent, "Service ID: %s", fastly.ToValue(&definition.ServiceID))
+		text.Indent(out, indent, "Name: %s", fastly.ToValue(&definition.Name))
+		text.Indent(out, indent, "Source: %s", fastly.ToValue(&definition.Source))
```
[the remark this answers is not in the document]
Comment (jsonroy-fastly) [reply pkg/commands/alerts/common.go:16]: That was a mistake. I kind of mixed up the actual `definitions` and the structs that we use with the go-fastly client (eg. `CreateAlertDefinitionInput`) that all uses pointers as fields.

File: pkg/commands/alerts/list.go (hunk 4 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/list.go:77]: Sounds good, we have some hard limits in the API that we cannot go over so I will add this suggested snippet.

File: pkg/commands/alerts/doc.go (hunk 3 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/doc.go:1]: Done

File: pkg/commands/alerts/create.go (hunk 1 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/create.go:27]: Done

File: pkg/commands/alerts/list_history.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	c.CmdClause.Flag("cursor", "Pagination cursor (Use 'next_cursor' value from list output)").Action(c.cursor.Set).StringVar(&c.cursor.Value)
+	c.CmdClause.Flag("limit", "Maximum number of items to list").Action(c.limit.Set).IntVar(&c.limit.Value)
+	c.CmdClause.Flag("sort", "Sort by one of the following [start]").Action(c.sort.Set).StringVar(&c.sort.Value)
+	c.CmdClause.Flag("order", "Sort by one of the following [asc, desc]").Action(c.order.Set).StringVar(&c.order.Value)
+	c.CmdClause.Flag("status", "Status of the history record [active, resolved]").Action(c.status.Set).StringVar(&c.status.Value)
+	c.CmdClause.Flag("after", "After filter history record that either started or ended after a specific date").Action(c.after.Set).StringVar(&c.after.Value)
+	c.CmdClause.Flag("before", "Before filter history record that either started or ended before a specific date").Action(c.before.Set).StringVar(&c.before.Value)
+	c.CmdClause.Flag("definition-id", "Unique identifier of the definition").Action(c.definitionID.Set).StringVar(&c.definitionID.Value)
+	c.CmdClause.Flag(argparser.FlagServiceIDName, "ServiceID of the definition").Action(c.serviceID.Set).StringVar(&c.serviceID.Value)
```
Comment (Integralist) [inline pkg/commands/alerts/list_history.go:33]: Sorted order...

```suggestion
	c.CmdClause.Flag("after", "After filter history record that either started or ended after a specific date").Action(c.after.Set).StringVar(&c.after.Value)
	c.CmdClause.Flag("before", "Before filter history record that either started or ended before a specific date").Action(c.before.Set).StringVar(&c.before.Value)
	c.CmdClause.Flag("cursor", "Pagination cursor (Use 'next_cursor' value from list output)").Action(c.cursor.Set).StringVar(&c.cursor.Value)
	c.CmdClause.Flag("definition-id", "Unique identifier of the definition").Action(c.definitionID.Set).StringVar(&c.definitionID.Value)
	c.RegisterFlagBool(c.JSONFlag()) // --json
	c.CmdClause.Flag("limit", "Maximum number of items to list").Action(c.limit.Set).IntVar(&c.limit.Value)
	c.CmdClause.Flag("order", "Sort by one of the following [asc, desc]").Action(c.order.Set).StringVar(&c.order.Value)
  c.CmdClause.Flag(argparser.FlagServiceIDName, "ServiceID of the definition").Action(c.serviceID.Set).StringVar(&c.serviceID.Value)
	c.CmdClause.Flag("sort", "Sort by one of the following [start]").Action(c.sort.Set).StringVar(&c.sort.Value)
	c.CmdClause.Flag("status", "Status of the history record [active, resolved]").Action(c.status.Set).StringVar(&c.status.Value)
```
```

File: pkg/commands/alerts/list.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	c.CmdClause = parent.Command("list", "List Alerts")
+
+	// Optional.
+	c.RegisterFlagBool(c.JSONFlag()) // --json
+	c.CmdClause.Flag("cursor", "Pagination cursor (Use 'next_cursor' value from list output)").Action(c.cursor.Set).StringVar(&c.cursor.Value)
+	c.CmdClause.Flag("limit", "Maximum number of items to list").Action(c.limit.Set).IntVar(&c.limit.Value)
+	c.CmdClause.Flag("name", "Name of the definition").Action(c.definitionName.Set).StringVar(&c.definitionName.Value)
+	c.CmdClause.Flag("sort", "Sort by one of the following [name, created_at, updated_at]").Action(c.sort.Set).StringVar(&c.sort.Value)
+	c.CmdClause.Flag("order", "Sort by one of the following [asc, desc]").Action(c.order.Set).StringVar(&c.order.Value)
+	c.CmdClause.Flag(argparser.FlagServiceIDName, "ServiceID of the definition").Action(c.serviceID.Set).StringVar(&c.serviceID.Value)
```
Comment (Integralist) [inline pkg/commands/alerts/list.go:30]: Sorted order...

```suggestion
	c.CmdClause.Flag("cursor", "Pagination cursor (Use 'next_cursor' value from list output)").Action(c.cursor.Set).StringVar(&c.cursor.Value)
	c.RegisterFlagBool(c.JSONFlag()) // --json
	c.CmdClause.Flag("limit", "Maximum number of items to list").Action(c.limit.Set).IntVar(&c.limit.Value)
	c.CmdClause.Flag("name", "Name of the definition").Action(c.definitionName.Set).StringVar(&c.definitionName.Value)
	c.CmdClause.Flag("order", "Sort by one of the following [asc, desc]").Action(c.order.Set).StringVar(&c.order.Value)
	c.CmdClause.Flag(argparser.FlagServiceIDName, "ServiceID of the definition").Action(c.serviceID.Set).StringVar(&c.serviceID.Value)
	c.CmdClause.Flag("sort", "Sort by one of the following [name, created_at, updated_at]").Action(c.sort.Set).StringVar(&c.sort.Value)
```

File: pkg/commands/alerts/update.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	c.CmdClause.Flag("source", "Source to get the metric from").Required().StringVar(&c.source)
+	c.CmdClause.Flag("type", "Type of evaluation strategy, one of [all_above_threshold, above_threshold, below_threshold, percent_absolute, percent_decrease, percent_increase]").Required().StringVar(&c.eType)
+	c.CmdClause.Flag("period", "Period of time to evaluate, one of [2m, 3m, 5m, 15m, 30m]").Required().StringVar(&c.period)
+	c.CmdClause.Flag("threshold", "Threshold use to compare during evaluation").Required().Float64Var(&c.threshold)
+
+	// Optional.
+	c.RegisterFlagBool(c.JSONFlag()) // --json
+	c.CmdClause.Flag("ignoreBelow", "IgnoreBelow use to discard data that are below this threshold").Action(c.ignoreBelow.Set).Float64Var(&c.ignoreBelow.Value)
+	c.CmdClause.Flag("dimensions", "Dimensions to filter on origins or domains").Action(c.dimensions.Set).StringsVar(&c.dimensions.Value)
+	c.CmdClause.Flag("integrations", "Integrations used to notify when an alert is firing or resolving").Action(c.integrations.Set).StringsVar(&c.integrations.Value)
```
Comment (Integralist) [inline pkg/commands/alerts/update.go:34]: Sorted order...

```suggestion
	c.CmdClause.Flag("dimensions", "Dimensions to filter on origins or domains").Action(c.dimensions.Set).StringsVar(&c.dimensions.Value)
	c.CmdClause.Flag("ignoreBelow", "IgnoreBelow use to discard data that are below this threshold").Action(c.ignoreBelow.Set).Float64Var(&c.ignoreBelow.Value)
	c.CmdClause.Flag("integrations", "Integrations used to notify when an alert is firing or resolving").Action(c.integrations.Set).StringsVar(&c.integrations.Value)
	c.RegisterFlagBool(c.JSONFlag()) // --json
```

File: pkg/commands/alerts/update.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	c.CmdClause = parent.Command("update", "Update Alerts")
+	c.CmdClause.Flag("id", "Alphanumeric string identifying an Alert definition").Required().StringVar(&c.definitionID)
+	c.CmdClause.Flag("name", "Name of the alert definition").Required().StringVar(&c.name)
+	c.CmdClause.Flag("description", "Description of the alert definition").Required().StringVar(&c.description)
+	c.CmdClause.Flag("metric", "Metric to alert on").Required().StringVar(&c.metric)
+	c.CmdClause.Flag("source", "Source to get the metric from").Required().StringVar(&c.source)
+	c.CmdClause.Flag("type", "Type of evaluation strategy, one of [all_above_threshold, above_threshold, below_threshold, percent_absolute, percent_decrease, percent_increase]").Required().StringVar(&c.eType)
+	c.CmdClause.Flag("period", "Period of time to evaluate, one of [2m, 3m, 5m, 15m, 30m]").Required().StringVar(&c.period)
+	c.CmdClause.Flag("threshold", "Threshold use to compare during evaluation").Required().Float64Var(&c.threshold)
```
Comment (Integralist) [inline pkg/commands/alerts/update.go:28]: Sorted order...

```suggestion
	c.CmdClause.Flag("description", "Description of the alert definition").Required().StringVar(&c.description)
	c.CmdClause.Flag("id", "Alphanumeric string identifying an Alert definition").Required().StringVar(&c.definitionID)
	c.CmdClause.Flag("metric", "Metric to alert on").Required().StringVar(&c.metric)
	c.CmdClause.Flag("name", "Name of the alert definition").Required().StringVar(&c.name)
	c.CmdClause.Flag("period", "Period of time to evaluate, one of [2m, 3m, 5m, 15m, 30m]").Required().StringVar(&c.period)
	c.CmdClause.Flag("source", "Source to get the metric from").Required().StringVar(&c.source)
	c.CmdClause.Flag("threshold", "Threshold use to compare during evaluation").Required().Float64Var(&c.threshold)
	c.CmdClause.Flag("type", "Type of evaluation strategy, one of [all_above_threshold, above_threshold, below_threshold, percent_absolute, percent_decrease, percent_increase]").Required().StringVar(&c.eType)
```

**Comment by jsonroy-fastly** [comment]
@Integralist There is one commit that I haven't pushed yet. Trying to resolve a enum panic

**Comment by jsonroy-fastly** [comment]
Another heads up I will be changing the description of fields to align to what we have on reference/api.

File: pkg/commands/alerts/create.go (hunk 2 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/create.go]: Done

File: pkg/commands/alerts/create.go (hunk 1 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/create.go]: It is now pushed

File: pkg/commands/alerts/create.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+		printVerbose(out, nil, definitions)
+	} else {
+		printSummary(out, nil, definitions)
+	}
+	return nil
+}
+
+// constructInput transforms values parsed from CLI flags into an object to be used by the API client library.
+func (c *CreateCommand) constructInput() *fastly.CreateAlertDefinitionInput {
+	eetype := c.eType
```
Comment (Integralist) [inline pkg/commands/alerts/create.go:87]: Looks like this variable is only used in one place so is an unnecessary indirection (i.e. extra mental override to track where it's used). I'd suggest just referencing `c.eType` directly on line 91 below.

File: pkg/commands/alerts/common.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+			start,
+			end,
+		)
+	}
+	t.Print()
+}
+
+// printVerbose displays the information returned from the API in a verbose
+// format.
+func printHistoryVerbose(out io.Writer, meta *fastly.AlertsMeta, history []*fastly.AlertHistory) {
```
Comment (Integralist) [inline pkg/commands/alerts/common.go:133]: As per the GitHub Actions linter, it looks like the `meta` argument can be removed as it's unused.

File: pkg/commands/alerts/common.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+		fmt.Fprintf(out, "Status: %s\n", history.Status)
+		fmt.Fprintf(out, "Start: %s\n", start)
+		fmt.Fprintf(out, "End: %s\n", end)
+		fmt.Fprintf(out, "\n")
+	}
+}
+
+// printSummary displays the information returned from the API in a summarised
+// format.
+func printHistorySummary(out io.Writer, meta *fastly.AlertsMeta, as []*fastly.AlertHistory) {
```
Comment (Integralist) [inline pkg/commands/alerts/common.go:114]: As per the GitHub Actions linter, it looks like the `meta` argument can be removed as it's unused.

File: pkg/commands/alerts/common.go [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+			threshold,
+			period,
+		)
+	}
+	t.Print()
+}
+
+// printVerbose displays the information returned from the API in a verbose
+// format.
+func printVerbose(out io.Writer, meta *fastly.AlertsMeta, as []*fastly.AlertDefinition) {
```
Comment (Integralist) [inline pkg/commands/alerts/common.go:91]: As per the GitHub Actions linter, it looks like the `meta` argument can be removed as it's unused.

File: pkg/commands/alerts/common.go [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+		}
+
+		text.Indent(out, indent, "Created at: %s", definition.CreatedAt)
+		text.Indent(out, indent, "Updated at: %s", definition.UpdatedAt)
+	}
+}
+
+// printSummary displays the information returned from the API in a summarised
+// format.
+func printSummary(out io.Writer, meta *fastly.AlertsMeta, as []*fastly.AlertDefinition) {
```
Comment (Integralist) [inline pkg/commands/alerts/common.go:68]: As per the GitHub Actions linter, it looks like the `meta` argument can be removed as it's unused.

**Review by Integralist — CHANGES_REQUESTED** [review]
LGTM. Just some final tweaks and then we're good to merge...

File: pkg/commands/alerts/common.go (hunk 15 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/common.go]: Done

File: pkg/commands/alerts/common.go (hunk 14 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/common.go]: Done

File: pkg/commands/alerts/common.go (hunk 13 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/common.go]: Done

File: pkg/commands/alerts/common.go (hunk 12 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/common.go]: Done

File: pkg/commands/alerts/create.go (hunk 11 above)
Comment (jsonroy-fastly) [reply pkg/commands/alerts/create.go]: Done

**Review by Integralist — APPROVED** [review]

**Pull request merged**